src/util/walkController.ts

```diff
@@ -0,0 +1,132 @@
+import { BaseTrie } from '..'
+import { FoundNodeFunction } from '../baseTrie'
+import { PrioritizedTaskExecutor } from '../prioritizedTaskExecutor'
+import { BranchNode, ExtensionNode, LeafNode, Nibbles, TrieNode } from '../trieNode'
+
+/**
+ * WalkController is an interface to control how the trie is being traversed.
+ */
+export default class WalkController {
+  readonly onNode: FoundNodeFunction
+  readonly taskExecutor: PrioritizedTaskExecutor
+  readonly trie: BaseTrie
+  private resolve: Function
+
+  /**
+   * Creates a new WalkController
+   * @param onNode - The `FoundNodeFunction` to call if a node is found
+   * @param trie - The `Trie` to walk on
+   */
+  private constructor(onNode: FoundNodeFunction, trie: BaseTrie, poolSize: number) {
+    this.onNode = onNode
+    this.taskExecutor = new PrioritizedTaskExecutor(poolSize)
+    this.trie = trie
+    this.resolve = () => {}
+  }
+
+  /**
+   * Async function to create and start a new walk over a trie.
+   * @param onNode - The `FoundNodeFunction to call if a node is found
+   * @param trie - The trie to walk on
+   * @param root - The root key to walk on
+   * @param poolSize - Task execution pool size to prevent OOM errors. Defaults to 500.
+   */
+  static async newWalk(
+    onNode: FoundNodeFunction,
+    trie: BaseTrie,
+    root: Buffer,
+    poolSize?: number
+  ): Promise<void> {
+    const strategy = new WalkController(onNode, trie, poolSize ?? 500)
+    await strategy.startWalk(root)
+  }
+
+  private async startWalk(root: Buffer): Promise<void> {
+    return await new Promise((resolve, reject) => {
+      this.resolve = resolve
+      this.trie
+        ._lookupNode(root)
+        .then((node) => {
+          this.processNode(root, node, [])
+        })
+        .catch((e) => {
+          reject(e)
+        })
+    })
```

Comment:
I think this would be better readable using `await` together with `try/catch`.

Reply:
I agree, but if I do that then the linter complains, because this is in a `new Promise`. I will rewrite it and add a `eslint-ignore` there.